src/openssl/mod.rs

```diff
@@ -310,20 +325,358 @@ impl ECVRF {
             }
         });
 
-        if let Some(pt) = point.as_mut() {
-            let mut new_pt = EcPoint::new(&self.group.as_ref())?;
-            new_pt.mul(
-                &self.group.as_ref(),
-                &pt,
-                &BigNum::from_slice(&[self.cofactor])?.as_ref(),
-                &self.bn_ctx,
-            )?;
-            *pt = new_pt;
+        if self.cofactor != 1 {
+            if let Some(pt) = point.as_mut() {
+                let mut new_pt = EcPoint::new(&self.group.as_ref())?;
+                new_pt.mul(
+                    &self.group.as_ref(),
+                    &pt,
+                    &BigNum::from_slice(&[self.cofactor])?.as_ref(),
+                    &self.bn_ctx,
+                )?;
+                *pt = new_pt;
+            }
         }
         // Return error if no valid point was found
         point.ok_or(Error::HashToPointError)
     }
 
+    // Utility function used in hash-to-curve methods.
+    // For security reasons, all field operations, comparisons, and assignments MUST be implemented in constant time
+    // (i.e., execution time MUST NOT depend on the values of the inputs), and without branching.
+
+    /// Function to move a field element conditionally.
+    /// Stated in [hash-to-curve-04](https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-04#section-4)
+    ///
+    ///
+    /// # Arguments
+    ///
+    /// * `a` - A field element
+    /// * `b` - A field element
+    /// * `c` - A boolean value
+    ///
+    /// # Returns
+    ///
+    /// * If c is False, CMOV returns a otherwise b
+    fn cmov(a: BigNum, b: BigNum, c: bool) -> BigNum {
+        // It has not been confirmed to have constant time
+        if c {
+            b
+        } else {
+            a
+        }
+    }
+
+    fn icmov(a: i32, b: i32, c: bool) -> i32 {
+        // It has not been confirmed to have constant time
+        if c {
+            b
+        } else {
+            a
+        }
+    }
+
+    /// Function to check whether the given field element is square or not.
+    /// Stated in [hash-to-curve-04](https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-04#section-4)
+    ///
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - A field element
+    ///
+    /// # Returns
+    ///
+    /// * Returns true whenever the value x is a square in the field given F otherwise false.
+
+    #[allow(clippy::wrong_self_convention)]
+    fn is_square(&mut self, x: &BigNum) -> Result<bool, Error> {
```

Comment:
When you move some repeated code to the exterior function, it'll be better if you put assertions about arguments. In here, it must be a field element over Fp, as you commented.

Reply:
Ah, thank you. That's what I'm missed.

Reply:
I made an assertion

Reply:
@HoOngEe It's not only for this function, but also other functions. Could you check once more?

Reply:
I added additional assertions to the functions that require a field element as an input.